synapse/storage/databases/main/beeper.py

```diff
@@ -47,50 +47,55 @@ async def beeper_preview_event_for_room_id_and_user_id(
     ) -> Optional[Tuple[str, int]]:
         def beeper_preview_txn(txn: LoggingTransaction) -> Optional[Tuple[str, int]]:
             sql = """
-            SELECT COALESCE(re.event_id, e.event_id), e.origin_server_ts
-            FROM events AS e
-            LEFT JOIN redactions as r
-                ON e.event_id = r.redacts
-            -- Join to relations from this event -> others to ignore edit events, we want to
-            -- skip events that replace others (caveat below) as they appear out-of order.
-            LEFT JOIN event_relations as er
-                ON e.event_id = er.event_id
-            -- Join to relations from other events -> this event to find any editing event, this
-            -- allows us to return the edited (newer) event, handling the case where the preview
-            -- event has been edited by the next event.
-            LEFT JOIN event_relations as err
-                ON e.event_id = err.relates_to_id AND err.relation_type = 'm.replace'
-            -- Join the replacing event
-            LEFT JOIN events as re
-                ON re.event_id = err.event_id
-            WHERE
-                e.stream_ordering <= ?
-                AND e.room_id = ?
-                AND (
-                    -- No relation or not replacements, so may include threads/etc
-                    er.relation_type IS NULL
-                    OR er.relation_type != 'm.replace'
-                )
-                AND r.redacts IS NULL
-                AND (
-                    e.type = 'm.room.message'
-                    OR e.type = 'm.room.encrypted'
-                    OR e.type = 'm.reaction'
-                )
-                AND CASE
-                    -- Only find non-redacted reactions to our own messages
-                    WHEN (e.type = 'm.reaction') THEN (
-                        SELECT ? = ee.sender AND ee.event_id NOT IN (
-                            SELECT redacts FROM redactions WHERE redacts = ee.event_id
-                        ) FROM events as ee
-                        WHERE ee.event_id = (
-                            SELECT eer.relates_to_id FROM event_relations AS eer
-                            WHERE eer.event_id = e.event_id
-                        )
+            WITH latest_event AS (
```

Comment:
Sorry Nick I got excited:
* Instead of just grabbing all event_relations, just look for replace edit relations and filter out those on `is_edit`
  * More stable in the case where an event has multiple relations
* We didn't need the "Join the replacing event" join
* One we've found the "latest event", we use a separate query to do edit handling to grab the latest content of the "latest event" based on edits. This allows us to handle the same event being edited multiple times.

Reply:
before
```
Limit  (cost=3.09..88.97 rows=1 width=48) (actual time=0.061..0.062 rows=1 loops=1)
  ->  Nested Loop Left Join  (cost=3.09..90950.01 rows=1059 width=48) (actual time=0.060..0.061 rows=1 loops=1)
        ->  Nested Loop Left Join  (cost=2.39..88505.29 rows=1059 width=104) (actual time=0.057..0.058 rows=1 loops=1)
              ->  Nested Loop Anti Join  (cost=1.82..65960.02 rows=1059 width=60) (actual time=0.048..0.049 rows=1 loops=1)
                    ->  Nested Loop Left Join  (cost=1.26..62674.84 rows=1113 width=60) (actual time=0.038..0.039 rows=1 loops=1)
                          Filter: ((er.relation_type IS NULL) OR (er.relation_type <> 'm.replace'::text))
                          ->  Index Scan Backward using events_room_stream on events e  (cost=0.70..56727.49 rows=1585 width=60) (actual time=0.026..0.026 rows=1 loops=1)
                                Index Cond: (room_id = '!HVqBdddiwPWjvZHHcJ:beeper.com'::text)
                                Filter: (((type = 'm.room.message'::text) OR (type = 'm.room.encrypted'::text) OR (type = 'm.reaction'::text)) AND CASE WHEN (type = 'm.reaction'::text) THEN (SubPlan 3) ELSE true END)
                                SubPlan 3
                                  ->  Index Scan using events_event_id_key on events ee  (cost=4.44..9.78 rows=1 width=1) (never executed)
                                        Index Cond: (event_id = $3)
                                        InitPlan 2 (returns $3)
                                          ->  Index Scan using event_relations_id on event_relations eer  (cost=0.56..3.74 rows=1 width=44) (never executed)
                                                Index Cond: (event_id = e.event_id)
                                        SubPlan 1
                                          ->  Index Only Scan using redactions_redacts on redactions  (cost=0.56..3.74 rows=1 width=44) (never executed)
                                                Index Cond: (redacts = ee.event_id)
                                                Heap Fetches: 0
                          ->  Index Scan using event_relations_id on event_relations er  (cost=0.56..3.74 rows=1 width=56) (actual time=0.010..0.010 rows=0 loops=1)
                                Index Cond: (event_id = e.event_id)
                    ->  Index Only Scan using redactions_redacts on redactions r  (cost=0.56..2.96 rows=1 width=44) (actual time=0.009..0.009 rows=0 loops=1)
                          Index Cond: (redacts = e.event_id)
                          Heap Fetches: 0
              ->  Index Scan using event_relations_relates on event_relations err  (cost=0.56..21.17 rows=12 width=88) (actual time=0.008..0.008 rows=0 loops=1)
                    Index Cond: ((relates_to_id = e.event_id) AND (relation_type = 'm.replace'::text))
        ->  Index Only Scan using events_event_id_key on events re  (cost=0.70..2.31 rows=1 width=44) (actual time=0.002..0.002 rows=0 loops=1)
              Index Cond: (event_id = err.event_id)
              Heap Fetches: 0
Planning Time: 0.857 ms
Execution Time: 0.110 ms

```

after
```
Nested Loop Left Join  (cost=104.24..104.29 rows=1 width=40) (actual time=0.079..0.081 rows=1 loops=1)
  Join Filter: (le.event_id = e.event_id)
  CTE latest_event
    ->  Limit  (cost=1.82..47.91 rows=1 width=60) (actual time=0.046..0.047 rows=1 loops=1)
          ->  Nested Loop Anti Join  (cost=1.82..67049.70 rows=1455 width=60) (actual time=0.046..0.047 rows=1 loops=1)
                ->  Nested Loop Anti Join  (cost=1.26..61405.83 rows=1508 width=60) (actual time=0.037..0.037 rows=1 loops=1)
                      ->  Index Scan Backward using events_room_stream on events e_1  (cost=0.70..56727.49 rows=1585 width=60) (actual time=0.024..0.025 rows=1 loops=1)
                            Index Cond: (room_id = '!HVqBdddiwPWjvZHHcJ:beeper.com'::text)
                            Filter: (((type = 'm.room.message'::text) OR (type = 'm.room.encrypted'::text) OR (type = 'm.reaction'::text)) AND CASE WHEN (type = 'm.reaction'::text) THEN (SubPlan 3) ELSE true END)
                            SubPlan 3
                              ->  Index Scan using events_event_id_key on events ee  (cost=4.44..9.78 rows=1 width=1) (never executed)
                                    Index Cond: (event_id = $3)
                                    InitPlan 2 (returns $3)
                                      ->  Index Scan using event_relations_id on event_relations eer  (cost=0.56..3.74 rows=1 width=44) (never executed)
                                            Index Cond: (event_id = e_1.event_id)
                                    SubPlan 1
                                      ->  Index Only Scan using redactions_redacts on redactions  (cost=0.56..3.74 rows=1 width=44) (never executed)
                                            Index Cond: (redacts = ee.event_id)
                                            Heap Fetches: 0
                      ->  Index Only Scan using redactions_redacts on redactions r  (cost=0.56..2.96 rows=1 width=44) (actual time=0.010..0.010 rows=0 loops=1)
                            Index Cond: (redacts = e_1.event_id)
                            Heap Fetches: 0
                ->  Index Scan using event_relations_id on event_relations is_edit  (cost=0.56..3.74 rows=1 width=44) (actual time=0.008..0.008 rows=0 loops=1)
                      Index Cond: (event_id = e_1.event_id)
                      Filter: (relation_type = 'm.replace'::text)
  ->  CTE Scan on latest_event le  (cost=0.00..0.02 rows=1 width=40) (actual time=0.049..0.049 rows=1 loops=1)
  ->  Limit  (cost=56.33..56.34 rows=1 width=84) (actual time=0.028..0.029 rows=1 loops=1)
        ->  Sort  (cost=56.33..56.36 rows=12 width=84) (actual time=0.028..0.028 rows=1 loops=1)
              Sort Key: e_replacement.origin_server_ts DESC
              Sort Method: quicksort  Memory: 25kB
              ->  Nested Loop Left Join  (cost=1.26..56.27 rows=12 width=84) (actual time=0.022..0.023 rows=1 loops=1)
                    ->  Nested Loop Left Join  (cost=0.56..16.75 rows=12 width=76) (actual time=0.013..0.014 rows=1 loops=1)
                          ->  CTE Scan on latest_event e  (cost=0.00..0.02 rows=1 width=32) (actual time=0.001..0.001 rows=1 loops=1)
                          ->  Index Only Scan using event_relations_relates on event_relations er  (cost=0.56..16.61 rows=12 width=44) (actual time=0.012..0.012 rows=0 loops=1)
                                Index Cond: ((relates_to_id = e.event_id) AND (relation_type = 'm.replace'::text))
                                Heap Fetches: 0
                    ->  Index Scan using events_event_id_key on events e_replacement  (cost=0.70..3.29 rows=1 width=52) (actual time=0.008..0.008 rows=0 loops=1)
                          Index Cond: (event_id = er.relates_to_id)
Planning Time: 0.661 ms
Execution Time: 0.151 ms
```